packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnGrouping.ts

```diff
@@ -0,0 +1,177 @@
+  }
+  const rep: UnwrappedGroupingModel = {};
+  const { groupId, children } = columnGroupNode;
+  children.forEach((child) => {
+    const unwrappedSubTree = recurrentUnwrapGroupingColumnModel(child, [...parents, groupId]);
+    Object.entries(unwrappedSubTree).forEach(([key, value]) => {
+      if (rep[key] !== undefined) {
+        throw new Error(
+          [
+            `MUI DataGrid - column grouping: duplicated field`,
+            `column field ${key} occurres two times in the grouping model:`,
+            `- ${rep[key].join(' > ')}`,
+            `- ${value.join(' > ')}`,
+          ].join('\n'),
+        );
+      }
+      rep[key] = value;
+    });
+  });
+  return rep;
+};
+
+export const unwrapGroupingColumnModel = (
+  columnGroupingModel?: GridColumnGroupingModel,
+): UnwrappedGroupingModel => {
+  if (!columnGroupingModel) {
+    return {};
+  }
+
+  const rep: UnwrappedGroupingModel = {};
+
+  columnGroupingModel.forEach((columnGroupNode) => {
+    const unwrappedSubTree = recurrentUnwrapGroupingColumnModel(columnGroupNode, []);
+    Object.entries(unwrappedSubTree).forEach(([key, value]) => {
+      if (rep[key] !== undefined) {
+        throw new Error(
+          [
+            `MUI: column grouping has duplicated field.`,
+            `column field ${key} occurres two times in the grouping model:`,
+            `- ${rep[key].join(' > ')}`,
+            `- ${value.join(' > ')}`,
+          ].join('\n'),
+        );
+      }
+      rep[key] = value;
+    });
+  });
+
+  return rep;
+};
+
+const createGroupLookup = (columnGroupingModel: GridColumnNode[]): GridColumnGroupLookup => {
+  let groupLookup: GridColumnGroupLookup = {};
+
+  columnGroupingModel.forEach((node) => {
+    if (isLeaf(node)) {
+      return;
+    }
+    const { groupId, children, ...other } = node;
+    if (!groupId) {
+      throw new Error(
+        'MUI-DataGrid: an element of the columnGroupingModel does not have either `field` or `groupId`',
+      );
+    }
+    if (!children) {
+      console.warn(`MUI-DataGrid: group groupId=${groupId} has no children. `);
+    }
+    const groupParam = { ...other, groupId };
+    const subTreeLookup = createGroupLookup(children);
+    if (subTreeLookup[groupId] !== undefined || groupLookup[groupId] !== undefined) {
+      throw new Error(`MUI: The groupId ${groupId} is used multiple times`);
+    }
+    groupLookup = { ...groupLookup, ...subTreeLookup, [groupId]: groupParam };
+  });
+
+  return { ...groupLookup };
+};
+export const columnGroupsStateInitializer: GridStateInitializer<
+  Pick<DataGridProcessedProps, 'columnGroupingModel'>
+> = (state, props) => {
+  const groupLookup = createGroupLookup(props.columnGroupingModel ?? []);
+  return {
+    ...state,
+    columnGrouping: { lookup: groupLookup, groupCollapsedModel: {} },
+  };
+};
+
+/**
+ * @requires useGridColumns (method, event)
+ * @requires useGridParamsApi (method)
+ */
+export const useGridColumnGrouping = (
```

Comment:
I would prefer not to add this kind of feature insid of the `columns` folder but instead create a `columnGrouping` folder
To keep as much as possible 1 folder = 1 sub state = 1 feature